src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -3299,6 +3300,271 @@ tx_gateway_frontend::get_all_transactions() {
+              return stm->read_lock().then(
+                [&self, stm, tid](ss::basic_rwlock<>::holder unit) {
+                    return with(
+                             stm,
+                             tid,
+                             "get_tx",
+                             [&self, stm, tid]() {
+                                 return self.describe_tx(stm, tid);
+                             })
+                      .finally([u = std::move(unit)] {});
+                });
+          });
+      });
+}
+
+ss::future<result<tm_transaction, tx_errc>> tx_gateway_frontend::describe_tx(
+  ss::shared_ptr<tm_stm> stm, kafka::transactional_id tid) {
+    auto term_opt = co_await stm->sync();
+    if (!term_opt.has_value()) {
+        if (term_opt.error() == tm_stm::op_status::not_leader) {
+            co_return tx_errc::not_coordinator;
+        }
+        co_return tx_errc::coordinator_not_available;
+    }
+    auto term = term_opt.value();
+
+    auto tx_opt = co_await stm->get_tx(tid);
+    if (!tx_opt.has_value()) {
+        auto status = tx_opt.error();
+        tx_errc err = tx_errc::unknown_server_error;
+        if (status == tm_stm::op_status::not_leader) {
+            err = tx_errc::leader_not_found;
+        } else if (status == tm_stm::op_status::timeout) {
+            err = tx_errc::timeout;
+        } else if (status == tm_stm::op_status::not_found) {
+            err = tx_errc::tx_id_not_found;
+        }
+        co_return err;
+    }
+
+    auto tx = tx_opt.value();
+
+    if (
+      tx.status == tm_transaction::tx_status::tombstone
+      || tm_transaction::tx_status::preparing) {
+        // We can just ignore this statuses
+        co_return tx;
+    } else if (tx.status == tm_transaction::tx_status::ready) {
+        if (term != tx.etag) {
+            if (!is_fetch_tx_supported()) {
+                vlog(
+                  txlog.warn,
+                  "tm_stm_cache feature is not supported. The tx({}) has "
+                  "started "
+                  "on the previous term and we can not get current state",
+                  tx.id);
+                co_return tx_errc::invalid_txn_state;
+            }
+        }
+
+        auto r1 = co_await fetch_tx(tx.id, tx.etag);
+        if (!r1) {
+            vlog(
+              txlog.warn,
+              "can't find old tx:{} etag:{} during describe_tx call",
+              tx.id,
+              tx.etag);
+            // Return possibly stale data
+            co_return tx;
+        }
+
+        auto old_tx = r1.value();
+
+        if (old_tx.pid != tx.pid) {
+            vlog(
+              txlog.warn,
+              "fetched pid:{} doesn't match local tx's pid:{}",
+              old_tx.pid,
+              tx.pid);
+            co_return tx_errc::unknown_server_error;
+        }
+
+        if (
+          old_tx.status != tm_transaction::tx_status::ongoing
+          && old_tx.status != tm_transaction::tx_status::ready) {
```

Comment:
`re_register_producer` and `reset_tx_ready` set status to ready and the transition happens from multiple statuses (I haven't tracked them all but they include at least prepare) and if the transition returned an error but actually passed the `fetch_tx` method may return previous state and it isn't limited only to ongoing & ready.

`re_register_producer` increments `tx_seq` (`tx.tx_seq += 1`); please update `reset_tx_ready` to increment it too and then use the following check here:

```
if (old_tx.status == tm_transaction::tx_status::ongoing) {
  if (old_tx.pid != tx.pid) {
    // add logging
    co_return tx_errc::unknown_server_error;
  }
} else if (old_tx.status == tm_transaction::tx_status::ready) {
  if (old_tx.tx_seq > tx.tx_seq) {
    // add logging
    co_return tx_errc::unknown_server_error;
  }
  old_tx = tx;
} else if (old_tx.tx_seq < tx.tx_seq) {
  old_tx = tx;
} else {
  // add logging
  co_return tx_errc::unknown_server_error;
}
```